0010-R-components-healthcheck.md

```diff
@@ -15,6 +15,16 @@ There are many components in Dapr which don't yet implement Ping.
 Ping is not mandatory to be implemented by Components, which is the correct behavior, as it could lead to false positives.
 For this components health-check, components not implementing Ping will be omitted out.
 
+## Use-Case
+If a mandatory* component fails at the start-up, Dapr will terminate or will move to some non-workable state like CrashLoopBackoff etc., so `healthz` API or any other API can't be used.
+
+After Dapr has started, if any Mandatory component fails, this healthcheck can be used to determine what component has failed and accordingly some steps acam be undertaken.
```

Comment:
> After Dapr has started, if any Mandatory component fails, this healthcheck can be used to determine what component has failed and accordingly some steps can be undertaken.

**During a period of time where the component health check is failing, what will be the effect on the sidecars operation?**

Will it be the same effect as with `App Health Check` i.e.

[Taken from the App Healthcheck Docs](https://docs.dapr.io/developing-applications/building-blocks/observability/app-health/)

> When it detects a failure in the app’s health, Dapr stops accepting new work on behalf of the application by:

- Unsubscribing from all pub/sub subscriptions
- Stopping all input bindings
- Short-circuiting all service-invocation requests, which terminate in the Dapr runtime and are not forwarded to the application

> These changes are meant to be temporary, and Dapr resumes normal operations once it detects that the application is responsive again.

Reply:
**Just my initial take**

If a component health check is failed, the sidecar **should** go into the same state as with a failed `App Health Check`.

This helps to keeps a consistent model for both types of downstream health check, which then makes it easier to reason about the behaviour of the sidecar during the period of time where one or more various probes/checks are failing.

Reply:
In case of App health, Dapr sidecar need to stop accepting events to process ( i.e. why subscriptions, input bindings or service invocation requests need to be stopped), as App itself is not Healthy and can't process these events. Do, Dapr sidecar doesn't know what to do with these events.

In case of a mandatory component health being reported as unhealthy, some features or All features of this component would be already un-usable. So, App can use this piece of information in a quick way to report back this status to 1. either some automated downstream to fix this issue or 2. Devops, which may consider some manual intervention.

But, here in case of a mandatory component being unhealthy, Dapr itself will not do any other operation. Here, rather App can decide to stop sending/receiving events via this component until it comes back.

Reply:
From my perspective, if the dapr sidecar knows that one or more mandatory components are not healthy, it makes little sense to allow the App to be invoked via PubSub/Service Invocation etc (I see this as inviting a preventable failure to occur)

Which is the same principle of App Health Checks -- if you know the App (or some downstream) is not healthy, don't invoke the App.

I absolutely see that some of those components may not always be needed, depending on the code path that is taken, so to enact the same behaviour as a failing App Health Check may seem heavy handed / aggressive.

However, I prefer an aggressive health check strategy -- "the sidecar is only as healthy as its weakest link" :)

---

The positive side effect of all of this is it will encourage operators to ensure that Components are `scoped` accordingly to the Apps that depend on them, rather than having no scope and Components being applied to every App.